src/dune_rules/pkg_rules.ml

```diff
@@ -1658,35 +1595,34 @@ module Compiler_dependency = struct
     | Some ocaml ->
       let+ toolchain_pkg = Resolve.resolve db context ocaml in
       (match toolchain_pkg with
+       | `Toolchain toolchain_version -> Toolchain toolchain_version
        | `System_provided -> System_provided
-       | `Inside_lock_dir pkg -> Pkg pkg)
+       | `Inside_lock_dir pkg -> Inside_lock_dir pkg)
   ;;
 
   let toolchain_version = function
+    | Toolchain toolchain_version -> Some toolchain_version
+    | Inside_lock_dir pkg ->
+      let module Toolchain = Dune_pkg.Toolchain in
```

Comment:
I would probably put this on the toplevel, since in the Dune codebase we rarely do local module bindings. Minor point.